### PR TITLE
Cache Loan User data from CRM

### PR DESCRIPTION
**Pull request by pawelpwc**

### CODE REVIEW

File: src/HE.InvestmentLoans.BusinessLogic/User/Repositories/LoanUserRepository.cs [hunk 1]
```diff
@@ -8,25 +9,43 @@ namespace HE.InvestmentLoans.BusinessLogic.User.Repositories;
...
 
         var resp = (invln_getcontactroleResponse)_serviceClient.Execute(req);
         if (resp.invln_portalroles != null)
         {
-            return JsonSerializer.Deserialize<ContactRolesDto>(resp.invln_portalroles);
+            var roles = JsonSerializer.Deserialize<ContactRolesDto>(resp.invln_portalroles);
+
+            if (roles != null)
+            {
+                _redisService.ObjectSet(redisKey, roles, 30);
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.BusinessLogic/User/Repositories/LoanUserRepository.cs:45]: I think that you don't have to pass hardcoded 30 (minutes) of expiration time. Lets use default value from the config.

**Review by lukasz-ciura-pwc — APPROVED** [review]

File: src/HE.InvestmentLoans.Common/Services/Interfaces/IRedisService.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+namespace HE.InvestmentLoans.Common.Services.Interfaces;
+
+public interface IRedisService
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.Common/Services/Interfaces/IRedisService.cs:3]: `IRedisService` -> `ICacheService` later we could have few implementation of cache services.

File: src/HE.InvestmentLoans.Common/Services/Interfaces/ICacheService.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+namespace HE.InvestmentLoans.Common.Services.Interfaces;
+
+public interface ICacheService
+{
+    public bool Exists(string key);
+
+    public T? ObjectGet<T>(string key);
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.Common/Services/Interfaces/ICacheService.cs:7]: Could we replace ObjectGet and ObjectSet method with one which return cache value if exists or make real call if cache is empty:
```csharp
public T GetCache(string key, Func<T> getValueFunc); 
```

Thanks to that using of cache will be super easy.

We can have quick call about it.

File: src/HE.InvestmentLoans.Common/HE.InvestmentLoans.Common.csproj [hunk 4]
```diff
@@ -20,6 +20,8 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.Common/HE.InvestmentLoans.Common.csproj:23]: Use `System.Text.Json` instead of `Newtonsoft.Json`

**Review by dawid-pawulski — APPROVED** [review]

File: src/HE.InvestmentLoans.Common/Models/App/CacheConfig.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+namespace HE.InvestmentLoans.Common.Models.App;
+
+public class CacheConfig : ICacheConfig
+{
+    public string RedisConnectionString { get; set; }
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.Common/Models/App/CacheConfig.cs:5]: `RedisConnectionString` -> `ConnectionString`?
```suggestion
    public string RedisConnectionString { get; set; }
```

File: src/HE.InvestmentLoans.Common/Services/MemoryCacheService.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        if (_memoryCache.TryGetValue(key, out T cacheValue))
+        {
+            return cacheValue;
+        }
+
+        var value = loadValue();
+
+        if (value != null)
+        {
+            var cacheEntryOptions = new MemoryCacheEntryOptions().SetSlidingExpiration(TimeSpan.FromSeconds(_config.ExpireMinutes));
```
Comment (lukasz-ciura-pwc) [inline src/HE.InvestmentLoans.Common/Services/MemoryCacheService.cs:30]: `FromSeconds` -> `FromMinutes` ?

**Review by lukasz-ciura-pwc — APPROVED** [review]

File: src/HE.InvestmentLoans.Common/Models/App/CacheConfig.cs (hunk 5 above)
Comment (lukasz-ciura-pwc) [reply src/HE.InvestmentLoans.Common/Models/App/CacheConfig.cs:5]: Or we could have a separate Config for only Redis.

**Pull request merged**